### PR TITLE
cosmos v0.44 support

### PR DESCRIPTION
**Pull request by ctrl-Felix**
add cosmos v0.44 support

### CODE REVIEW

**Review by czarcas7ic — COMMENTED** [review]
This fixed pushing txs to the new endpoint, tysm for this

**Comment by czarcas7ic** [comment]
But also to note, I had to change one line under get_pushable command

```
        return json.dumps(
           # {"jsonrpc": "2.0", "id": 1, "method": self._sync_mode, "params": {"tx": tx_b64}}
           { "tx_bytes": tx_b64, "mode": "BROADCAST_MODE_SYNC"}
```

File: README.md [hunk 1]
```diff
@@ -126,15 +126,19 @@ tx.add_transfer(
...
 
 
-# Optionally submit the transaction using your preferred method.
-# This example uses the httpx library.
-import httpx
+# Submit the transaction
+rpc_url = "https://rpc.cosmos.network/"
+
+# Method 1
+r = tx.broadcast(url=rpc_url)
```
Comment (hukkin) [inline README.md:135]: I don't really want this API or the `requests` dependency. Cosmospy is not a networking library.

**Comment by ctrl-Felix** [comment]
> I had to change one line under get_pushable command

Wasn't it working like it was there? Because I submitted transactions like this

File: README.md [hunk 2]
```diff
@@ -126,15 +126,19 @@ tx.add_transfer(
...
 
-api_base_url = "https://node.atomscan.com"
-httpx.post(api_base_url + "/txs", data=pushable_tx)
+# Method 2
+import requests
+
+pushable_tx = tx.get_pushable()
+r = requests.post(rpc_url, data=pushable_tx)
```
Comment (hukkin) [inline README.md:141]: Let's stick to `httpx` here.

File: pyproject.toml [hunk 3]
```diff
@@ -27,6 +27,10 @@ dependencies = [
     "mnemonic >=0.19",
     "hdwallets >=0.1.0",
     "typing-extensions >=3.7.4; python_version < '3.8'",
+    "requests==2.26.0",
+    "protobuf >= 3.19.1",
+    "google >= 3.0.0",
```
Comment (hukkin) [inline pyproject.toml:32]: where is `google` dependency used? Do we really need it?

File: src/cosmospy/_transaction.py [hunk 4]
```diff
@@ -3,12 +3,17 @@
 import base64
 import hashlib
 import json
-from typing import Any
 
 import ecdsa
+import requests
 
+from cosmospy._typing import SyncMode
```
Comment (hukkin) [inline src/cosmospy/_transaction.py:10]: `typing` module should remain public API. (no underscore prefix)

File: src/cosmospy/_transaction.py [hunk 5]
```diff
@@ -44,64 +49,86 @@ def __init__(
...
+
+    def _get_fee(self):
+        _fee = coin.Coin()
+        _fee.amount = str(self._fee)
+        _fee.denom = self._fee_denom
+        return _fee
+
+    def _get_pubkey(self):
+        pubkey_bytes = privkey_to_pubkey(self._privkey)
+        _pubkey = pubkey.PubKey()
```
Comment (hukkin) [inline src/cosmospy/_transaction.py:113]: There shouldn't be a reason to use private variables inside the function scope (applies to other methods too)

File: src/cosmospy/_typing.py [hunk 6]
```diff
@@ -1,14 +1,15 @@
 import sys
 
+print(sys.version_info)
```
Comment (hukkin) [inline src/cosmospy/_typing.py:3]: This should be removed

File: src/cosmospy/_typing.py [hunk 7]
```diff
@@ -1,14 +1,15 @@
...
 if sys.version_info < (3, 8):
     from typing_extensions import Literal, TypedDict
 else:
     from typing import Literal, TypedDict
 
 
-# Valid transaction broadcast modes for the `POST /txs` endpoint of the
+# Valid transaction broadcast modes for the TX endpoint of the
```
Comment (hukkin) [inline src/cosmospy/_typing.py:10]: `POST /cosmos/tx/v1beta1/txs` endpoint

**Review by hukkin — CHANGES_REQUESTED** [review]
Thanks a lot!

I'm not yet able to have a good look unfortunately but here's my first drive-by review.

A couple questions:
- where did you get the `.proto` files?
- are all the `.proto` files strictly needed? Is this the smallest set that works?
- what command did you run to generate `.py` files from the `.proto` files?

**Comment by hukkin** [comment]
> > I had to change one line under get_pushable command

> Wasn't it working like it was there? Because I submitted transactions like this

I'm fairly sure @czarcas7ic is correct here. Based on [swagger docs](https://v1.cosmos.network/rpc/v0.42.6) of the `POST /cosmos/tx/v1beta1/txs` endpoint.

**Comment by czarcas7ic** [comment]
> > I had to change one line under get_pushable command
> 
> Wasn't it working like it was there? Because I submitted transactions like this

Also see https://docs.cosmos.network/master/core/proto-docs.html#cosmos-tx-v1beta1-service-proto

**Comment by ctrl-Felix** [comment]
> > > I had to change one line under get_pushable command
> 
> > Wasn't it working like it was there? Because I submitted transactions like this
> 
> I'm fairly sure @czarcas7ic is correct here. Based on [swagger docs](https://v1.cosmos.network/rpc/v0.42.6) of the `POST /cosmos/tx/v1beta1/txs` endpoint.

I think it just depends if you use the API or RPC. If you want to use it with the RPC endpoint it works. That's why I changed the endpoint to https://rpc.cosmos.network and not the api endpoint.

If you want to submit it to the API you need the change @czarcas7ic made.

It would be interesting to know if there is a performance difference

**Comment by ctrl-Felix** [comment]
> Thanks a lot!
> 
> I'm not yet able to have a good look unfortunately but here's my first drive-by review.
> 
> A couple questions:
> 
>     * where did you get the `.proto` files?
> 
>     * are all the `.proto` files strictly needed? Is this the smallest set that works?
> 
>     * what command did you run to generate `.py` files from the `.proto` files?

I took the proto files from cyberpy but I think those are the sames they also used in their js library. I will check if they are all needed. This version is just a really fast adaption because I wanted to provide a fix for everyone using cosmospy (also for me I wasn't able to vroadcast transactions in python anymore)

I took the generated python files also from cybai but you can generate those files with tools like https://github.com/dropbox/mypy-protobuf, I will setup a proper way to generate them also in case of changements. 

Regarding the google dep it is needed as everything relies on google.protobuf and if google or protobuf aren't installed you get the error ```ModuleNotFoundError: No module named 'google.protobuf'```

Also the current tests will need to be recreated to work with the new update.

**Comment by hukkin** [comment]
> I took the proto files from cyberpy but I think those are the sames they also used in their js library. I will check if they are all needed. 

Thanks. Would be good to know what is `cyberpy`s source for these files instead of blindly copying them.

> Regarding the google dep it is needed as everything relies on google.protobuf and if google or protobuf aren't installed you get the error ModuleNotFoundError: No module named 'google.protobuf'

The `google.protobuf` namespace should be available solely from `protobuf` package. See https://github.com/protocolbuffers/protobuf/tree/master/python where the source code is under `google/protobuf/`.

**Comment by ctrl-Felix** [comment]
Okay, I have some news regarding the proto files

The proto files are coming from cosmos (ex: https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/base/v1beta1/coin.proto => https://github.com/ctrl-Felix/cosmospy/blob/sdkv0.44/src/cosmospy/proto/coin.proto) 

There are then cli utilities to migrate those protos to python code.

I saw some proto files have been updated on cosmos or have more content. I will update them on my repo too.

**Comment by ctrl-Felix** [comment]
> The `google.protobuf` namespace should be available solely from `protobuf` package. See https://github.com/protocolbuffers/protobuf/tree/master/python where the source code is under `google/protobuf/`.

I am confused too but it needs both. Check out this thread: https://stackoverflow.com/questions/38680593/importerror-no-module-named-google-protobuf

There is no explanation to this

**Comment by hukkin** [comment]
> I am confused too but it needs both. Check out this thread: https://stackoverflow.com/questions/38680593/importerror-no-module-named-google-protobuf

Did you try removing the `google` dependency? We don't do `import google`, the autogenerated files do `from google.protobuf import *`. I don't think we need the dependency.

**Comment by hukkin** [comment]
> Okay, I have some news regarding the proto files

I think this may be the best source for the proto files https://buf.build/cosmos/cosmos-sdk/tree

What we need is
- the minimal set of proto files that we need for `cosmospy` to work (downloaded from the resource above)
- the command to autogenerate Python

I cannot accept a PR with autogenerated files that I can not verify/reproduce myself.

**Comment by ctrl-Felix** [comment]
I will close the pr anyways and redo it with a new version where the protos are generated from the official ones. This PR is probbaly more a dirty quick fix.

**Comment by ctrl-Felix** [comment]
> > I am confused too but it needs both. Check out this thread: https://stackoverflow.com/questions/38680593/importerror-no-module-named-google-protobuf
> 
> Did you try removing the `google` dependency? We don't do `import google`, the autogenerated files do `from google.protobuf import *`. I don't think we need the dependency.

ah yeah, you're right. Google is the one which isn't needed.

**Comment by hukkin** [comment]
BTW, here's the docs for compiling Python from `.proto` https://developers.google.com/protocol-buffers/docs/pythontutorial#compiling-your-protocol-buffers

Here's a link to download the (currently) latest `protoc` binaries https://github.com/protocolbuffers/protobuf/releases/tag/v3.19.3

**Comment by ctrl-Felix** [comment]
I have some problems with converting the proto files to python files. AS Example:
Coin.proto looks like this on cyberpy https://github.com/SaveTheAles/cyberpy/blob/master/cyberpy/interfaces/coin_pb2.py
My converted version looks like this https://github.com/ctrl-Felix/cosmospy/blob/master/src/cosmospy/generated/coin_pb2.py

**Pull request closed without merging**